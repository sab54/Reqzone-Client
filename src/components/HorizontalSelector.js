/**
 * HorizontalSelector.js
 *
 * What This Component Does:
 * - Renders a horizontally scrollable list of selectable "chips".
 * - Highlights the currently selected item via theme-driven colors.
 * - Invokes `onSelect(item)` when a chip is pressed.
 *
 * Props:
 * - data: Array<any>                     // Items to render. Each item must at least provide a `label` field for display.
 * - selected: any                        // The currently selected item.
 * - onSelect: (item: any) => void        // Callback when an item is pressed.
 * - theme?: {
 *     primary?: string,                  // Background color for selected chip.
 *     surface?: string,                  // Background color for unselected chip.
 *     border?: string,                   // Border color for all chips.
 *     text?: string,                     // Text color for unselected chip.
 *     buttonPrimaryText?: string,        // Text color for selected chip.
 *   }
 * - isEqual?: (a: any, b: any) => boolean  // Equality fn to determine selection (default strict equality).
 * - renderIcon?: (item: any, isSelected: boolean) => React.Node  // Optional icon renderer per chip.
 * - itemKey?: (item: any) => string | number                     // Key extractor for React lists (defaults to the item itself).
 *
 * Behavior:
 * - Selected chip uses `{ backgroundColor: theme.primary, color: theme.buttonPrimaryText, fontWeight: '600' }`.
 * - Unselected chip uses `{ backgroundColor: theme.surface, color: theme.text, fontWeight: '500' }`.
 * - Scrolls horizontally without a scrollbar; spacing and rounded corners applied for chip look.
 *
 * Notes:
 * - No testIDs are required; tests should query by visible label text and inspect styles.
 * - Ensure Poppins font is loaded in the app if using custom fonts.
 *
 * Author: Sunidhi Abhange
 */

import React from 'react';
import {
    View,
    Text,
    ScrollView,
    TouchableOpacity,
    StyleSheet,
} from 'react-native';

const HorizontalSelector = ({
    data = [],
    selected,
    onSelect,
    theme = {},
    isEqual = (a, b) => a === b,
    renderIcon,
    itemKey = (item) => item,
}) => {
    return (
        <ScrollView
            horizontal
            showsHorizontalScrollIndicator={false}
            contentContainerStyle={styles.scrollContainer}
        >
            {data.map((item) => {
                const isSelected = isEqual(item, selected);
                return (
                    <TouchableOpacity
                        key={itemKey(item)}
                        onPress={() => onSelect(item)}
                        style={[
                            styles.item,
                            {
                                backgroundColor: isSelected
                                    ? theme.primary || '#4B7BE5'
                                    : theme.surface || '#f0f0f0',
                                borderColor: theme.border || '#ddd',
                            },
                        ]}
                    >
                        {renderIcon?.(item, isSelected)}
                        <Text
                            style={[
                                styles.label,
                                {
                                    color: isSelected
                                        ? theme.buttonPrimaryText || '#fff'
                                        : theme.text || '#333',
                                    fontWeight: isSelected ? '600' : '500',
                                },
                            ]}
                        >
                            {item.label}
                        </Text>
                    </TouchableOpacity>
                );
            })}
        </ScrollView>
    );
};

const styles = StyleSheet.create({
    scrollContainer: {
        flexDirection: 'row',
        paddingBottom: 8,
    },
    item: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingHorizontal: 12,
        paddingVertical: 6,
        borderRadius: 20,
        borderWidth: 1,
        marginRight: 8,
    },
    label: {
        marginLeft: 6,
        fontSize: 14,
        fontFamily: 'Poppins',
    },
});

export default HorizontalSelector;
