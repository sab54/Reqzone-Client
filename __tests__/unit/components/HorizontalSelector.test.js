/**
 * HorizontalSelector.test.js
 *
 * What This Test File Covers:
 *
 * 1. Renders items
 *    - Ensures all labels from `data` are visible and optional icon renderer is used.
 *
 * 2. Selection styles (text only)
 *    - Selected item text uses theme.buttonPrimaryText & fontWeight '600'.
 *    - Unselected item text uses theme.text & fontWeight '500'.
 *
 * 3. onSelect callback
 *    - Pressing a label calls `onSelect` with the correct item.
 *
 * 4. Snapshot
 *    - Captures stable output for regression checks.
 *
 * Notes:
 * - No component changes (no testIDs). Queries rely on visible text.
 */

import React from 'react';
import { Text } from 'react-native'; // ✅ needed for renderIcon
import { render, fireEvent } from '@testing-library/react-native';
import HorizontalSelector from 'src/components/HorizontalSelector';

const theme = {
  primary: '#0044ff',
  surface: '#eeeeee',
  border: '#cccccc',
  text: '#111111',
  buttonPrimaryText: '#ffffff',
};

const data = [
  { id: 1, label: 'One' },
  { id: 2, label: 'Two' },
  { id: 3, label: 'Three' },
];

describe('HorizontalSelector', () => {
  it('renders all items and optional icons', () => {
    const { getByText } = render(
      <HorizontalSelector
        data={data}
        selected={data[2]}
        onSelect={() => {}}
        theme={theme}
        itemKey={(it) => it.id}
        renderIcon={(it, isSelected) => <Text>{isSelected ? '★' : '☆'}</Text>}
      />
    );

    expect(getByText('One')).toBeTruthy();
    expect(getByText('Two')).toBeTruthy();
    expect(getByText('Three')).toBeTruthy();
    expect(getByText('★')).toBeTruthy(); // selected icon
  });

  it('applies text color and fontWeight based on selection', () => {
    const { getByText } = render(
      <HorizontalSelector
        data={data}
        selected={data[1]} // "Two" selected
        onSelect={() => {}}
        theme={theme}
        itemKey={(it) => it.id}
      />
    );

    const selectedText = getByText('Two');
    const unselectedText = getByText('One');

    expect(selectedText.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ color: theme.buttonPrimaryText }),
        expect.objectContaining({ fontWeight: '600' }),
      ])
    );

    expect(unselectedText.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ color: theme.text }),
        expect.objectContaining({ fontWeight: '500' }),
      ])
    );
  });

  it('calls onSelect with the pressed item', () => {
    const onSelect = jest.fn();
    const { getByText } = render(
      <HorizontalSelector
        data={data}
        selected={data[0]}
        onSelect={onSelect}
        theme={theme}
        itemKey={(it) => it.id}
      />
    );

    fireEvent.press(getByText('Three'));
    expect(onSelect).toHaveBeenCalledWith(data[2]);
  });

  it('matches snapshot', () => {
    const tree = render(
      <HorizontalSelector
        data={data}
        selected={data[0]}
        onSelect={() => {}}
        theme={theme}
        itemKey={(it) => it.id}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});