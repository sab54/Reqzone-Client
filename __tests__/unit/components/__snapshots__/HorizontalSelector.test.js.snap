// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizontalSelector matches snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "flexDirection": "row",
      "paddingBottom": 8,
    }
  }
  horizontal={true}
  showsHorizontalScrollIndicator={false}
>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#0044ff",
          "borderColor": "#cccccc",
          "borderRadius": 20,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginRight": 8,
          "opacity": 1,
          "paddingHorizontal": 12,
          "paddingVertical": 6,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Poppins",
              "fontSize": 14,
              "marginLeft": 6,
            },
            {
              "color": "#ffffff",
              "fontWeight": "600",
            },
          ]
        }
      >
        One
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#eeeeee",
          "borderColor": "#cccccc",
          "borderRadius": 20,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginRight": 8,
          "opacity": 1,
          "paddingHorizontal": 12,
          "paddingVertical": 6,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Poppins",
              "fontSize": 14,
              "marginLeft": 6,
            },
            {
              "color": "#111111",
              "fontWeight": "500",
            },
          ]
        }
      >
        Two
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#eeeeee",
          "borderColor": "#cccccc",
          "borderRadius": 20,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginRight": 8,
          "opacity": 1,
          "paddingHorizontal": 12,
          "paddingVertical": 6,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Poppins",
              "fontSize": 14,
              "marginLeft": 6,
            },
            {
              "color": "#111111",
              "fontWeight": "500",
            },
          ]
        }
      >
        Three
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
